098-using-image-volumes-to-improve-extensibility-of-Strimzi-operands.md

```diff
@@ -0,0 +1,320 @@
+In that case, Kubernetes will ignore the image volume and its volume mount and create the Pod without it (tested with Minikube on Kubernetes 1.31 and 1.32, and with OpenShift 4.18).
+As a result, the plugin will be missing and the user has to choose another way how to add it.
+
+In case the image volumes are enabled in Kubernetes but not supported by the container runtime, the container will end up in the `CreateContainerError` state with the following message:
+
+```yaml
+  status:
+    state:
+      waiting:
+        message: 'Error response from daemon: invalid volume specification: '':/mnt/echo-sink-plugin/:ro'''
+        reason: CreateContainerError
+```
+
+In such case, the user will need to edit the related Strimzi custom resource, remove the parts using the image volumes and apply the changes.
+Afterwards, the Strimzi operator will recover and recreate the Pod with a valid configuration.
+
+Use of a feature gate to prevent the use of image volumes on unsupported cluster was considered.
+But since it is expected that we will be seeing Kubernetes clusters which do not support image volumes for a very long time, the feature gate which would be for a long time in alpha phase and would need to be always enabled does not seem to make sense.
+
+### Testing strategy
+
+The functionality from this proposal will be initially tested only using unit tests.
+This should be sufficient in terms of test coverage given we only create the right Pod definition and the rest is up to Kubernetes.
+Introducing a system test right now would mean that such a test would be executable only on very specific Kubernetes versions with specific configurations and would not add much value.
+System tests using image volumes might be added at a later stage once the image volume support reaches more Kubernetes clusters.
+
+## Out of scope
+
+The image volumes have the potential to simplify how we add our own plugins to our Apache Kafka container images and how we structure them.
+We might for example have just a simple Strimzi base image and have the operator mount the different components into it as image volumes on demand:
+* The supported Kafka versions
+* 3rd party libraries (i.e. plugins shipped by Strimzi out of the box)
+* Cruise Control and Kafka Exporter
+
+This would make it easier to handle base image CVEs but also CVEs in the different components.
+It would also make sure that software that might contain CVEs will not be actually in the container image unless it s actually needed / used.
+
+However, to be able to propose and implement something like this, we first need all Kubernetes clusters supported by Strimzi to also enable and support the ImageVolume feature gate.
+And it might take several years until we get there.
+Until then, the image volumes can be used only for the _optional_ Strimzi features that have some other alternatives.
```

Comment:
shall we list the optional components that we propose to include for now? Maybe this information should be more prominent than out of scope? An In Scope section?

Reply:
I tried to reword it a bit. What I was trying to say here is that the features added by this proposal are only optional for the user. So for them we do not mind if the image volumes are supported only somewhere. But the other possibilities mentioned in the `Out of scope` section would be required and if we implemented them, Strimzi would not work at all in environments without the image volume support. So that is why we cannot consider these out of scope features right now.